HCP Terraform Docs (#1664)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/cloud-docs/registry/test/dynamic-credentials/aws.mdx

Before you begin, decide whether to use separate IAM roles for module tests or reuse workspace roles. Separate roles provide better isolation and allow you to grant different permissions for testing versus production workloads.

## Requirements

You need the following to configure dynamic credentials for AWS in module testing:

- An AWS account with permissions to create and configure IAM roles and OIDC providers
- An HCP Terraform organization with module testing enabled
- A private registry module with [testing configured](/terraform/cloud-docs/registry/test)
- Permissions in HCP Terraform to modify registry module test settings

## Configure AWS

### Create an IAM OIDC provider

If you haven't already configured an OIDC identity provider for HCP Terraform in your AWS account, create one:

1. Sign in to the AWS Management Console and navigate to **IAM > Identity providers**
2. Click **Add provider**
3. Select **OpenID Connect** as the provider type
4. Enter the provider URL:
   - For HCP Terraform: `https://app.terraform.io`

>    - For Terraform Enterprise: `https://<TFE_HOSTNAME>` (replace with your TFE hostname)

Reviewer: ```suggestion
   - For Terraform Enterprise: `https://<your-terraform-enterprise-hostname>`. 
```
Avoid "TFE" per the [corporate style guide](https://docs.google.com/document/d/1MRvGd6tS5JkIwl_GssbyExkMJqOXKeUE00kSEtFi8m8/edit?tab=t.0#heading=h.95uakj2jl8yp).

8. Name the role (e.g., `terraform-module-test-role`)
9. Click **Create role**

### Configure the trust policy

After creating the role, configure its trust policy to accept tokens from your module tests. The trust policy controls which HCP Terraform organizations and modules can assume the role.

1. Open the IAM role you created
2. Navigate to the **Trust relationships** tab
3. Click **Edit trust policy**
4. Update the trust policy to include module test-specific conditions:

<CodeBlockConfig hideClipboard>

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Federated": "arn:aws:iam::123456789012:oidc-provider/app.terraform.io"
      },
      "Action": "sts:AssumeRoleWithWebIdentity",
      "Condition": {
        "StringEquals": {
          "app.terraform.io:aud": "aws.workload.identity"
        },
        "StringLike": {
          "app.terraform.io:sub": "organization:my-org:module:*:operation:test_run"
        }
      }
    }
  ]
}
```

</CodeBlockConfig>

Replace the following values:
- `123456789012`: Your AWS account ID
- `app.terraform.io`: Your HCP Terraform or Terraform Enterprise hostname

> - `my-org`: Your organization's name (found in organization settings)

Reviewer: ```suggestion
- `my-org`: Your organization's name. You can get the name from your organization settings page.
```

3. Click **Configuration** or **Settings**
4. In the **Dynamic Credentials** section, toggle **Enable dynamic credentials**
5. Select **AWS** as the provider
6. Enter the IAM role ARN (e.g., `arn:aws:iam::123456789012:role/terraform-module-test-role`)
7. (Optional) Set a custom audience if you configured a different audience in your OIDC provider
8. Click **Save configuration**

### Configure through the API

You can also configure dynamic credentials using the [Test Configuration API](/terraform/cloud-docs/api-docs/private-registry/test-configuration):

```bash
curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request PATCH \
  --data @payload.json \
  https://app.terraform.io/api/v2/registry-modules/:registry_name/:namespace/:name/:provider/test-configuration
```

With the following `payload.json`:

```json
{
  "data": {
    "type": "test-configurations",
    "attributes": {
      "oidc-enabled": true,
      "oidc-provider": "aws",
      "oidc-configuration": {
        "role-arn": "arn:aws:iam::123456789012:role/terraform-module-test-role",
        "audience": "aws.workload.identity"
      }
    }
  }
}
```

> ## Environment variables

Reviewer: ```suggestion
### Environment variables
```
I'm a little confused about this section. Is it a step in the process of configuring HCP Terraform? Even if it's not, it still seems to be a related subtopic and should be grouped under that heading.

When you configure dynamic credentials for AWS, HCP Terraform automatically sets the following environment variables in your module test runs:

| Variable | Description | Example Value |
|----------|-------------|---------------|
| `TFC_AWS_PROVIDER_AUTH` | Signals the AWS provider to use OIDC authentication | `true` |
| `TFC_AWS_RUN_ROLE_ARN` | The IAM role ARN to assume | `arn:aws:iam::123456789012:role/terraform-module-test-role` |
| `TFC_AWS_WORKLOAD_IDENTITY_AUDIENCE` | The audience claim for OIDC tokens | `aws.workload.identity` |
| `TFC_OIDC_ISSUER_URL` | The OIDC issuer URL | `https://app.terraform.io` |
| `TFC_OIDC_AUDIENCE` | The OIDC audience | `aws.workload.identity` |

The AWS Terraform provider (version 4.67.0 and later) automatically uses these environment variables to authenticate with AWS. You don't need to configure the provider block explicitly.

## Trust policy examples

> 

Reviewer: ```suggestion
The following examples show common trust policy configurations.
```
Feel free to rephrase. I'm not sure if there is such a thing as a common configuration pattern.

### Read-only testing

For modules that only need to validate configurations without creating resources:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "ec2:Describe*",
        "s3:List*",
        "s3:Get*",
        "iam:Get*",
        "iam:List*"
      ],
      "Resource": "*"
    }
  ]
}
```

### Resource creation for testing

For modules that create test resources, limit permissions to specific resource types and use resource tags:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "ec2:RunInstances",
        "ec2:TerminateInstances",
        "ec2:CreateTags",
        "ec2:Describe*"
      ],
      "Resource": "*",
      "Condition": {
        "StringEquals": {
          "aws:RequestedRegion": "us-east-1",
          "ec2:CreateAction": "RunInstances"
        }
      }
    },
    {
      "Effect": "Allow",
      "Action": [
        "ec2:TerminateInstances"
      ],
      "Resource": "arn:aws:ec2:us-east-1:123456789012:instance/*",
      "Condition": {
        "StringEquals": {
          "ec2:ResourceTag/Testing": "true"
        }
      }
    }
  ]
}
```

> -> **Tip**: Use resource tags like `Testing: true` or `ManagedBy: terraform-test` to identify and scope permissions for test resources.

Reviewer: ```suggestion
<Tip>

Use resource tags, such as `Testing: true` or `ManagedBy: terraform-test`, to identify and scope permissions for test resources.

</Tip>
```

> ## Verify the configuration

Reviewer: ```suggestion
## Verify the configuration
```
We should move this section to just below the Configure HCP Terraform section so that all of the core steps are grouped together. The H2s in the sidecar nav should read:

Requirements
Configure AWS
Configure HCP Terraform (includes Environment variables)
Verify configuration
Permission recommendations
Trust policy examples
Troubleshooting
Next steps

1. Navigate to your module in HCP Terraform
2. Go to the **Tests** tab
3. Click **Start Test Run** or trigger a test through your VCS
4. Monitor the test run logs for successful authentication

If authentication succeeds, you'll see log output indicating the AWS provider authenticated using the OIDC token. If it fails, check the [troubleshooting](#troubleshooting) section.

## Troubleshooting

### Error: "No valid credential sources found"

This error indicates the AWS provider cannot find credentials. Check:

1. The AWS provider version is 4.67.0 or later
2. Dynamic credentials are enabled in the test configuration
3. The role ARN is correct in the test configuration
4. The `TFC_AWS_PROVIDER_AUTH` environment variable is set to `true`

### Error: "Not authorized to perform sts:AssumeRoleWithWebIdentity"

This error means the IAM role's trust policy rejected the OIDC token. Verify:

1. The OIDC provider exists in your AWS account
2. The trust policy includes the correct issuer URL (app.terraform.io or your TFE hostname)
3. The subject claim pattern matches your organization name and module, ending with `:operation:test_run`

> 4. The audience in the trust policy matches the configured audience

Reviewer: ```suggestion
1. The OIDC provider exists in your AWS account.
1. The trust policy includes the correct issuer URL, such as `app.terraform.io` for HCP Terraform or your Terraform Enterprise hostname.
1. The subject claim pattern matches your organization name and module and ends with `:operation:test_run`.
1. The audience in the trust policy matches the configured audience.
```

### Error: "Token has expired"

> OIDC tokens expire after the configured TTL (default: 10 minutes). If your tests run longer:

Reviewer: ```suggestion
OIDC tokens expire after the configured TTL. The default TTL is 10 minutes. If your tests run longer, perform the following steps:
```
